drivers/input/input_renesas_ra_ctsu.c

```diff
@@ -0,0 +1,676 @@
+#include <soc.h>
+
+#include <zephyr/drivers/clock_control/renesas_ra_cgc.h>
+#include <zephyr/input/input_renesas_ra_ctsu.h>
+#include "rm_touch.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(renesas_ra_touch, CONFIG_INPUT_LOG_LEVEL);
+
+struct renesas_ra_ctsu_cfg {
+	const struct gpio_dt_spec tscap_pin;
+	const struct pinctrl_dev_config *pcfg;
+	const struct device *clock;
+	const struct clock_control_ra_subsys_cfg clock_subsys;
+	void (*make_thread)(const struct device *dev);
+	void (*irq_config)(void);
+};
+
+struct renesas_ra_ctsu_data {
+	struct k_sem scanning;
+	struct k_queue scan_q;
+	struct k_thread thread_data;
+};
+
+struct renesas_ra_ctsu_group_cfg {
+	const struct device *ctsu_dev;
+	size_t num_button;
+	size_t num_slider;
+	size_t num_wheel;
+};
+
+struct renesas_ra_ctsu_device_cb {
+	const struct device *dev;
+	void (*device_cb)(const struct device *dev, void *data);
+};
+
+struct renesas_ra_ctsu_group_data {
+	const struct device *dev;
+	/* FSP Touch data */
+	struct st_touch_instance touch_instance;
+#ifndef CONFIG_INPUT_RENESAS_RA_QE_TOUCH_CFG
+	struct st_touch_instance_ctrl touch_ctrl;
+	struct st_touch_cfg touch_cfg;
+	/* FSP CTSU data */
+	struct st_ctsu_instance ctsu_instance;
+	struct st_ctsu_instance_ctrl ctsu_ctrl;
+	struct st_ctsu_cfg ctsu_cfg;
+#endif /* CONFIG_INPUT_RENESAS_RA_QE_TOUCH_CFG */
+	/* Touch driver private data */
+	struct k_work reading_work;
+	struct k_timer sampling_timer;
+	/* Touch driver sample result */
+	uint64_t *p_button_status;
+	uint16_t *p_slider_position;
+	uint16_t *p_wheel_position;
+	/* Touch device callback data */
+	struct renesas_ra_ctsu_device_cb *p_button_cb;
+	struct renesas_ra_ctsu_device_cb *p_slider_cb;
+	struct renesas_ra_ctsu_device_cb *p_wheel_cb;
+};
+
+extern void ctsu_write_isr(void);
+extern void ctsu_read_isr(void);
+extern void ctsu_end_isr(void);
+
+struct ctsu_device_cfg {
+	const struct device *group_dev;
+	uint16_t event_code;
+};
+
+struct ctsu_scan_msg {
+	void *reserved; /* first word of queue data item reserved for the kernel */
+	struct st_touch_instance *p_instance;
+};
+
+SYS_MEM_BLOCKS_DEFINE_STATIC(scan_msg_allocator, sizeof(struct ctsu_scan_msg),
+			     CONFIG_INPUT_RENESAS_RA_CTSU_MSG_MEM_BLOCK_SIZE, sizeof(uint32_t));
+
+static void renesas_ra_callback_adapter(touch_callback_args_t *p_args)
+{
+	const struct device *dev = p_args->p_context;
+	const struct renesas_ra_ctsu_group_cfg *cfg = dev->config;
+	const struct device *ctsu_dev = cfg->ctsu_dev;
+	struct renesas_ra_ctsu_data *ctsu_data = ctsu_dev->data;
+	struct renesas_ra_ctsu_group_data *data = dev->data;
+
+	if (p_args->event == CTSU_EVENT_SCAN_COMPLETE) {
+		k_work_submit(&data->reading_work);
+	}
+
+	k_sem_give(&ctsu_data->scanning);
+}
+
+#define K_POLLING       K_MSEC(CONFIG_INPUT_RENESAS_RA_CTSU_POLLING_INTERVAL_MS)
+#define K_STABILIZATION K_USEC(CONFIG_INPUT_RENESAS_RA_CTSU_STABILIZATION_TIME_US)
+
+static ALWAYS_INLINE void renesas_ra_ctsu_drv_handler(void *arg0)
```

Comment:
why always inline? I'd imagine the compiler is smart enough to optimize it by itself

the argument should probably just be a`const struct device *` no point keep casting it back and forth